drivers/include/adc_ng.h

```diff
@@ -0,0 +1,333 @@
+
+/**
+ * @brief   Chance the ADC MUX settings to switch to a different ADC channel
+ *
+ * @param[in]       adc     Running ADC device to change the MUX settings of
+ * @param[in]       chan    Channel to switch to
+ *
+ * @retval  0               Success
+ * @retval  -ENXIO          Invalid channel given in @p chan
+ * @retval  -ENOTSUP        Cannot switch to channel @p chan using the
+ *                          current configuration of the ADC
+ * @retval  -EIO            The ADC is not powered up and configured
+ * @retval  <0              Other error (see device driver doc)
+ *
+ * @pre     The ADC is up and running (see @ref adc_ng_init). Otherwise
+ *          `-EIO` will be returned
+ * @post    Except for the ADC MUX, no ADC setting is changed by this
+ * @post    On success, the ADC MUX is set to @p chan
+ * @post    On failure, the ADC MUX setting is left unchanged
+ */
+static inline int adc_ng_mux(uint8_t adc, uint8_t chan)
+{
+    assert(adc < XFA_LEN(adc_ng_backend_t, adc_ng_backends));
+    const adc_ng_backend_t be = adc_ng_backends[adc];
+    return be.driver->mux(be.handle, chan);
+}
+
+/**
+ * @brief   Turn of the given ADC device
+ *
+ * @param[in]       adc     ADC device to turn off
+ */
+static inline void adc_ng_off(uint8_t adc)
+{
+    assert(adc < XFA_LEN(adc_ng_backend_t, adc_ng_backends));
+    const adc_ng_backend_t be = adc_ng_backends[adc];
+    if (be.driver->off) {
+        be.driver->off(be.handle);
+    }
+}
+
+/**
+ * @brief   Return the number of the highest single ended channel of the
+ *          given ADC
+ *
+ * @param[in]       adc     ADC to get the number of the highest single
+ *                          ended channel of
+ *
+ * @details All channel from zero to the returned value are single ended
+ *          channels
+ * @note    Some ADCs might have "holes" in the channel map, e.g. channels
+ *          0, 1 and 3 might be available, but channel 2 is not supported.
+ *          Checking for the return value `-ENXIO` in @ref adc_ng_init can be
+ *          used to detect these "holes".
+ *
+ * A single ended input is the "normal" ADC channel where a single input
+ * voltage is converted into an unsigned integer with the value 0 referring to
+ * an input voltage of 0 V and the highest value referring to an input voltage
+ * equal to the voltage reference (or higher).
+ */
+static inline uint8_t adc_ng_highest_single_ended_channel(uint8_t adc)
+{
+    assert(adc < XFA_LEN(adc_ng_backend_t, adc_ng_backends));
+    const adc_ng_backend_t be = adc_ng_backends[adc];
+    return be.driver->highest_single_ended_channel;
+}
+
+/**
+ * @name    Functions access to access the raw ADC output
+ *
+ * These functions return the binary value returned by the ADC, rather than
+ * meaningful physical units.
+ *
+ * @{
+ */
+
+/**
+ * @brief   Perform a single conversion using the specified ADC channel
+ *
+ * @param[in]       adc     ADC device to use
+ * @param[out]      dest    The result is stored here
```

Comment:
What format is `result`?
e.g. with a 12 bit ADC, will the max value be 4095 or 4294967295?

Reply:
anything but left alignment wouldn't make sense :)

so max value == int32 max and min value ==  int32 min, if not the value would depend on the chose resolution.

Reply:
but why is the return value int wont this get us into trouble for avr (int == int16)

Reply:
This intents to also work for external ADCs, e.g. 24 bit ADCs on 8 bit MCUs.

Reply:
oops i thought return is the returned value, but the return value is just an error code